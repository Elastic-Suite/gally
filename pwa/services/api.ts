import { apiUrl } from '~/constants'

export function fetchApi<T>(
  language: string,
  url: RequestInfo | URL,
  options: RequestInit = {}
): Promise<T> {
  if (process.env.NEXT_PUBLIC_LOCAL) {
    if (!String(url).startsWith('/')) {
      url = `/${url}`
    }
    if (!String(url).endsWith('.json') && !String(url).endsWith('.jsonld')) {
      url = `${url}.json`
    }
    url = `/mocks${url}`
  }
  url = `${apiUrl}${url}`
  return fetch(url, {
    ...options,
    headers: {
      ...options.headers,
      'Accept-Language': language,
    },
  }).then((response) => response.json())
}
