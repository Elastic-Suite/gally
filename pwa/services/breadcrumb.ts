import { IMenuChild } from '~/store'

export function getSlugArray(data: string[] | string): string[] {
  const newBreadCrumbData = [data[0]]
  for (let index = 0; index < data.length - 1; index++) {
    newBreadCrumbData.push(`${newBreadCrumbData[index]}_${data[index + 1]}`)
  }
  return newBreadCrumbData
}

export function findBreadcrumbLabel(
  findIndex: number,
  slug: string[],
  menu: IMenuChild[],
  deepIndex = 0
): string {
  if (typeof menu === 'object') {
    for (const menuChild in menu) {
      if (menu[menuChild].code === slug[deepIndex]) {
        if (findIndex === deepIndex) {
          return menu[menuChild].label
        }
        return findBreadcrumbLabel(
          findIndex,
          slug,
          menu[menuChild].children,
          deepIndex + 1
        )
      }
    }
  }
  return null
}
