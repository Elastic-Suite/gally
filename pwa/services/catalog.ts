import { ICatalog } from '~/types'

export function findDefaultCatalog(catalogsData: ICatalog[]): ICatalog | null {
  const defaultCatalog = catalogsData
    ? catalogsData.filter((catalog) =>
        catalog.localizedCatalogs.find((localizedCtl) => localizedCtl.isDefault)
      )[0]
    : null

  if (defaultCatalog) {
    defaultCatalog.localizedCatalogs = defaultCatalog.localizedCatalogs.filter(
      (localizedCtl) => localizedCtl.isDefault
    )
  }
  return defaultCatalog ? defaultCatalog : null
}
