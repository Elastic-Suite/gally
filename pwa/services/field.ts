import { IElasticSuiteProperty, IField } from '~/types'

function updateProperties(
  properties: IElasticSuiteProperty,
  key: string,
  value: number | boolean
): IElasticSuiteProperty {
  if (key === 'visible') {
    return {
      ...properties,
      visible: value as boolean,
    }
  } else if (key === 'editable') {
    return {
      ...properties,
      editable: value as boolean,
    }
  } else if (key === 'position') {
    return {
      ...properties,
      position: value as number,
    }
  }
  return properties
}

export function updatePropertiesAccordingToPath(
  field: IField,
  path: string
): IField {
  if (path.includes('admin/settings')) {
    path = 'settings_attribute'
  } else {
    path = path.replaceAll('/', '_').replace('_admin_', '')
  }
  if (field.elasticsuite?.context) {
    const [_, newPropertiesvalues] = Object.entries(field.elasticsuite?.context)
      .filter(([key, _]) => key === path)
      .flat()
    if (newPropertiesvalues) {
      let newProperties = field.elasticsuite
      // eslint-disable-next-line no-return-assign
      Object.entries(newPropertiesvalues).forEach(
        ([property, propertyValue]) =>
          (newProperties = updateProperties(
            newProperties,
            property,
            propertyValue
          ))
      )
      return {
        ...field,
        elasticsuite: newProperties,
      }
    }
  }
  return field
}
