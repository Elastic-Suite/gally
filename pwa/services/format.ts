import inflection from 'inflection'

export function firstLetterUppercase(item: string): string {
  return item[0].toUpperCase() + item.slice(1)
}

export function firstLetterLowercase(item: string): string {
  return item[0].toLowerCase() + item.slice(1)
}

export function humanize(label: string): string {
  return inflection.transform(label, ['underscore', 'humanize'])
}

export function getFieldLabelTranslationArgs(
  source: string,
  resource?: string
): [string, string] {
  const { sourceWithoutDigits, sourceSuffix } = getSourceParts(source)
  const defaultLabel = humanize(sourceSuffix.replace(/\./g, ' '))

  if (resource) {
    return [`resources.${resource}.fields.${sourceWithoutDigits}`, defaultLabel]
  }

  return [`fields.${sourceWithoutDigits}`, defaultLabel]
}

// Taken from https://github.com/marmelab/react-admin/blob/master/packages/ra-core/src/util/getFieldLabelTranslationArgs.ts
function getSourceParts(source: string): {
  sourceWithoutDigits: string
  sourceSuffix: string
} {
  const sourceWithoutDigits = source.replace(/\.\d+\./g, '.')
  const parts = source.split('.')
  let lastPartWithDigits
  parts.forEach((part, index) => {
    if (onlyDigits(part)) {
      lastPartWithDigits = index
    }
  })
  const sourceSuffix =
    lastPartWithDigits !== null && typeof lastPartWithDigits !== 'undefined'
      ? parts.slice(lastPartWithDigits + 1).join('.')
      : source

  return { sourceWithoutDigits, sourceSuffix }
}

// Taken from https://github.com/marmelab/react-admin/blob/master/packages/ra-core/src/util/getFieldLabelTranslationArgs.ts
function onlyDigits(s: string): boolean {
  for (let i = s.length - 1; i >= 0; i--) {
    const d = s.charCodeAt(i)
    if (d < 48 || d > 57) return false
  }
  return true
}
