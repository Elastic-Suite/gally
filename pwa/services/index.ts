export * from './api'
export * from './authProvider'
export * from './format'
export * from './hydra'
export * from './jsonld'
export * from './local'
export * from './table'
export * from './tests'
