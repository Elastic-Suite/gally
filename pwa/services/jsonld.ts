import { IDocs } from '~/store'
import { IFetch, IHydraPropertyTypeRef, IHydraSupportedProperty } from '~/types'

export function getJsonldClass(
  docs: IFetch<IDocs>,
  propertyClass: string
): IHydraSupportedProperty[] {
  if (!propertyClass.startsWith('#')) {
    propertyClass = `#${propertyClass}`
  }
  return docs.data.jsonld['hydra:supportedClass'].find(
    (classes) =>
      classes['@type'] === 'hydra:Class' && classes['@id'] === propertyClass
  )?.['hydra:supportedProperty']
}

export function getJsonldClassRef(
  docs: IFetch<IDocs>,
  apiProperty: IHydraPropertyTypeRef
): IHydraSupportedProperty[] {
  const [_, propertyClass] = /^#\/components\/schemas\/([^.]*)/.exec(
    apiProperty.$ref
  )
  if (propertyClass) {
    return getJsonldClass(docs, propertyClass)
  }
}

export function getJsonldProperty(
  hydraClass: IHydraSupportedProperty[],
  propertyName: string
): IHydraSupportedProperty {
  return hydraClass.find((property) => property['hydra:title'] === propertyName)
}

export function getJsonldLinkClass(
  docs: IFetch<IDocs>,
  propery: IHydraSupportedProperty
): IHydraSupportedProperty[] {
  if (propery['hydra:property']['@type'] === 'hydra:Link') {
    return getJsonldClass(docs, propery['hydra:property'].range)
  }
}
