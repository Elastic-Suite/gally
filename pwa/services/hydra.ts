import { Api, Field, Resource } from '@api-platform/api-doc-parser'

import { booleanRegexp, fieldIdRegexp } from '~/constants'
import {
  IHydraMember,
  IHydraResponse,
  IOptions,
  ISearchParameters,
} from '~/types'

import { firstLetterLowercase } from './format'

export function getResource(doc: Api, resourceName: string): Resource {
  return doc.resources?.find((resource) => resource.name === resourceName)
}

export function getReadableFieldName(property: string): string {
  const result = booleanRegexp.exec(property)
  if (result?.[1]) {
    return firstLetterLowercase(result[1])
  }
  if (property.endsWith('[]')) {
    return property.slice(0, -2)
  }
  return property
}

export function getFieldNameFromFieldId(id: string): string {
  const match = fieldIdRegexp.exec(id)
  return match?.[1]
}

export function getReadableField(resource: Resource, name: string): Field {
  name = getReadableFieldName(name).toLowerCase()
  return resource.readableFields.find((field) => {
    const fieldName = getFieldNameFromFieldId(field.id)
    return (
      (fieldName && fieldName.toLowerCase() === name) ||
      field.name.toLowerCase() === name
    )
  })
}

// See https://github.com/api-platform/admin/blob/main/src/hydra/schemaAnalyzer.ts
export function getFieldType(field: Field): string {
  switch (field.id) {
    case 'http://schema.org/identifier':
      return field.range === 'http://www.w3.org/2001/XMLSchema#integer'
        ? 'integer_id'
        : 'id'
    case 'http://schema.org/email':
      return 'email'
    case 'http://schema.org/url':
      return 'url'
    default:
  }

  if (field.embedded !== null && field.maxCardinality !== 1) {
    return 'array'
  }

  switch (field.range) {
    case 'http://www.w3.org/2001/XMLSchema#array':
      return 'array'
    case 'http://www.w3.org/2001/XMLSchema#integer':
      return 'integer'
    case 'http://www.w3.org/2001/XMLSchema#decimal':
    case 'http://www.w3.org/2001/XMLSchema#float':
      return 'float'
    case 'http://www.w3.org/2001/XMLSchema#boolean':
      return 'boolean'
    case 'http://www.w3.org/2001/XMLSchema#date':
      return 'date'
    case 'http://www.w3.org/2001/XMLSchema#dateTime':
      return 'dateTime'
    default:
      return 'text'
  }
}

export function getOptionsFromApi<T extends IHydraMember>(
  response: IHydraResponse<T>
): IOptions {
  return response['hydra:member'].map((member) => ({
    id: member.id,
    label: member['@id'],
    value: member.id,
  }))
}

export function castFieldParameter(
  field: Field,
  value: string | string[]
): string | number | boolean | (string | number | boolean)[] {
  if (value instanceof Array) {
    return value.map(
      (value) => castFieldParameter(field, value) as string | number | boolean
    )
  }
  if (field.reference) {
    return Number(value)
  }
  switch (field.type) {
    case 'integer':
      return Number(value)
    case 'boolean':
      return value !== 'true' && value !== 'false' ? null : value === 'true'
    default:
      return value
  }
}

export function isFieldValueValid(field: Field, value: unknown): boolean {
  if (value instanceof Array) {
    return value.every((value) => isFieldValueValid(field, value))
  }
  if (field.reference) {
    return typeof value === 'number' && !isNaN(value)
  }
  switch (field.type) {
    case 'integer':
      return typeof value === 'number' && !isNaN(value)
    case 'boolean':
      return typeof value === 'boolean'
    default:
      return typeof value === 'string'
  }
}

export function getFilterParameters(
  resource: Resource,
  parameters: ISearchParameters
): ISearchParameters {
  return Object.fromEntries(
    Object.entries(parameters).reduce((acc, [key, value]) => {
      const field = getReadableField(resource, key)
      if (field) {
        const fieldValue = castFieldParameter(field, value as string | string[])
        if (isFieldValueValid(field, fieldValue)) {
          acc.push([key, fieldValue])
        }
      }
      return acc
    }, [])
  )
}
