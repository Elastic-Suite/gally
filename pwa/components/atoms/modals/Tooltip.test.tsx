import Tooltips from './Tooltip'
import { renderWithProviders } from '~/services'

describe('TooltipLeft', () => {
  it('match snapshot', () => {
    const { container } = renderWithProviders(
      <Tooltips title="Hello World" placement="left">
        <span>Hello World</span>
      </Tooltips>
    )
    expect(container).toMatchSnapshot()
  })
})

describe('TooltipRight', () => {
  it('match snapshot', () => {
    const { container } = renderWithProviders(
      <Tooltips title="Hello World" placement="right">
        <span>Hello World</span>
      </Tooltips>
    )
    expect(container).toMatchSnapshot()
  })
})

describe('TooltipTop', () => {
  it('match snapshot', () => {
    const { container } = renderWithProviders(
      <Tooltips title="Hello World" placement="top">
        <span>Hello World</span>
      </Tooltips>
    )
    expect(container).toMatchSnapshot()
  })
})

describe('TooltipBottom', () => {
  it('match snapshot', () => {
    const { container } = renderWithProviders(
      <Tooltips title="Hello World" placement="bottom">
        <span>Hello World</span>
      </Tooltips>
    )
    expect(container).toMatchSnapshot()
  })
})
