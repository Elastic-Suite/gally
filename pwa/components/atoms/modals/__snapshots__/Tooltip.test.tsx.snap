// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TooltipBottom match snapshot 1`] = `
<div>
  <span
    aria-label="Hello World"
    class=""
    data-mui-internal-clone-element="true"
  >
    Hello World
  </span>
</div>
`;

exports[`TooltipLeft match snapshot 1`] = `
<div>
  <span
    aria-label="Hello World"
    class=""
    data-mui-internal-clone-element="true"
  >
    Hello World
  </span>
</div>
`;

exports[`TooltipRight match snapshot 1`] = `
<div>
  <span
    aria-label="Hello World"
    class=""
    data-mui-internal-clone-element="true"
  >
    Hello World
  </span>
</div>
`;

exports[`TooltipTop match snapshot 1`] = `
<div>
  <span
    aria-label="Hello World"
    class=""
    data-mui-internal-clone-element="true"
  >
    Hello World
  </span>
</div>
`;
